odoo/modules/module.py

```diff
@@ -465,22 +465,115 @@ def get_test_modules(module):
+
+    def getDescription(self, test):
+        """
+        Simplified version of TextTestResult getDescription
+        """
+        if isinstance(test, unittest.TestCase):
+            return "%s.%s" % (test.__class__.__qualname__, test._testMethodName)
+        return str(test)
+
+    def startTest(self, test):
+        super(OdooTestResult, self).startTest(test)
+        self.log(logging.INFO, 'Starting test %s' % self.getDescription(test), test=test)
+
+    def addError(self, test, err):
+        super(OdooTestResult, self).addError(test, err)
+        self.logError("ERROR", test, err)
+
+    def addFailure(self, test, err):
+        super(OdooTestResult, self).addFailure(test, err)
+        self.logError("FAIL", test, err)
+
+    def addSkip(self, test, reason):
+        super(OdooTestResult, self).addSkip(test, reason)
+        self.log(logging.INFO, 'skipped %s' % self.getDescription(test), test=test)
+
+    def addUnexpectedSuccess(self, test):
+        super(OdooTestResult, self).addUnexpectedSuccess(test)
+        self.log(logging.ERROR, 'unexpected success for %s' % self.getDescription(test), test=test)
+
+    def logError(self, flavour, test, error):
+        err = self._exc_info_to_string(error, test)
+        caller_infos = self.getErrorCallerInfo(error, test)
+        self.log(logging.INFO, '=' * 70, test, caller_infos=caller_infos)
+        self.log(logging.ERROR, "%s: %s\n%s" % (flavour, self.getDescription(test), err), test=test, caller_infos=caller_infos)
+
+    def getErrorCallerInfo(self, error, test):
+        """
+        :param error: A tupple (exctype, value, tb) as returned by sys.exc_info().
+        :param test: The start date
+        :returns: a tupple (fn, lno, func, sinfo) matching the logger findCaller format or None
+        """
+
+        # only test case should be executed in odoo, if not use default CallerInfo
+        if not isinstance(test, unittest.TestCase):#
             return
-        level = logging.ERROR if s.startswith(('ERROR', 'FAIL', 'Traceback')) else logging.INFO
-        self.logger.log(level, s)
+        _, _, error_traceback = error
+        lineno = 0
+
+        while error_traceback:
+            code = error_traceback.tb_frame.f_code
+            if code.co_name == test._testMethodName:
+                lineno = error_traceback.tb_lineno
+                filename = code.co_filename
+                method = test._testMethodName
+                infos = (filename, lineno, method, None)
+                return infos
+            error_traceback = error_traceback.tb_next
+
+
+class OdooTestRunner(object):
+    """A test runner class that displays results in in logger.
+    Simplified verison of TextTestRunner(
+    """
+
+    def run(self, test):
+        "Run the given test case or test suite."
+        result = OdooTestResult()
+        # unittest.signal.registerResult(result)
+        # import warnings
+        # with warnings.catch_warnings():
+        start_time = time.perf_counter()
```

Comment:
Our tests are unlikely to run fast enough that this is really useful. Even more so if we're only keeping milliseconds at formatting.

Reply:
This is to keep the same log as before, this is the TextTestRunner implementation

Reply:
>  this is the TextTestRunner implementation.

Ah indeed, it looks like the implementation was changed recently (on 3.6 it uses time.time): https://github.com/python/cpython/commit/8db5b54463118e5eb60cb3582e3108623f01f5df

Apparently it switched to perf_counter because it's monotonic rather than for performance reasons, so fair enough.